test/scale/scale_tester.go

```diff
@@ -0,0 +1,358 @@
+	// Get the list of available nodes for scheduling the pod
+	nodes := nodeutil.ListNodeNames(client)
+	if len(nodes) < 3 {
+		ginkgo.Skip("need at least 3 nodes to verify the test case. Current node count is %d", len(nodes))
+	}
+
+	ginkgo.By("Initiating statefulset setup")
+
+	statefulSets := []*testtypes.TestStatefulset{}
+
+	tStorageClass, storageCleanup := t.Pod.CreateStorageClass(client, namespace, t.CSIDriver, t.StorageClassParameters)
+	defer storageCleanup()
+
+	for i := 0; i < t.PodCount; i++ {
+		statefulSet, cleanup := t.Pod.SetupStatefulset(client, namespace, t.CSIDriver, schedulerName, 1, t.StorageClassParameters, &tStorageClass)
+		for j := range cleanup {
+			defer cleanup[j]()
+		}
+
+		statefulSets = append(statefulSets, statefulSet)
+	}
+
+	ginkgo.By("Completed statefulset setup")
+
+	ginkgo.By("Initiating statefulset deployment")
+	var wg sync.WaitGroup
+	start := time.Now()
+	for i := range statefulSets {
+		wg.Add(1)
+		go func(statefulset *testtypes.TestStatefulset) {
+			defer wg.Done()
+			defer ginkgo.GinkgoRecover()
+
+			statefulset.Create()
+		}(statefulSets[i])
+	}
+
+	ginkgo.By("Waiting for the pod for all the statefulsets to succeed")
+	wg.Wait()
+	framework.Logf("Time taken to complete deployment of statefulsets : %d", time.Since(start).Milliseconds())
+
+	ginkgo.By("All pods for the statefulsets are ready")
+	ginkgo.By("Initiating scaling down all the statefulsets")
+
+	for i := range statefulSets {
+		newScale := &scale.Scale{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      statefulSets[i].Statefulset.Name,
+				Namespace: statefulSets[i].Namespace.Name,
+			},
+			Spec: scale.ScaleSpec{
+				Replicas: int32(0),
+			},
+		}
+
+		_, err := client.AppsV1().StatefulSets(statefulSets[i].Namespace.Name).UpdateScale(context.TODO(), statefulSets[i].Statefulset.Name, newScale, metav1.UpdateOptions{})
+		framework.ExpectNoError(err)
+	}
+
+	ginkgo.By("Sleep 30mins waiting for statefulset update to complete")
+	time.Sleep(30 * time.Minute)
+
+	ginkgo.By("Reseting the scale for the statefulsets to 1")
+	start = time.Now()
+
+	for i := range statefulSets {
+		newScale := &scale.Scale{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      statefulSets[i].Statefulset.Name,
+				Namespace: statefulSets[i].Namespace.Name,
+			},
+			Spec: scale.ScaleSpec{
+				Replicas: int32(1),
+			},
+		}
+
+		wg.Add(1)
+		go func(ss *testtypes.TestStatefulset) {
+			defer wg.Done()
+			defer ginkgo.GinkgoRecover()
+
+			_, err := client.AppsV1().StatefulSets(ss.Namespace.Name).UpdateScale(context.TODO(), ss.Statefulset.Name, newScale, metav1.UpdateOptions{})
+			framework.ExpectNoError(err)
+			time.Sleep(30 * time.Second)
+			ss.WaitForPodReady()
+		}(statefulSets[i])
+	}
+
+	wg.Wait()
+	framework.Logf("Time taken to rescale the statefulsets : %d", time.Since(start).Milliseconds())
+}
+
+func (t *PodSchedulingOnFailoverMultiplePVScaleTest) Run(client clientset.Interface, namespace *v1.Namespace, schedulerName string) {
```

Comment:
This seems to be doing the same thing as the above test with the exception that the pod definition here has more than one volume. Why not reuse `PodSchedulingOnFailoverScaleTest`?

Reply:
Makes sense. Thanks for pointing this out.